src/error.rs

```diff
@@ -3,33 +3,114 @@
+  Error,
+  /// A fatal error
+  Fatal,
+}
 
-impl StructuredError {
-  /// Wrap around and own a raw xmllib2 error structure
-  pub fn from_raw(error: *mut bindings::_xmlError) -> Self {
-    Self(error)
+impl XmlErrorLevel {
+  /// Convert an xmlErrorLevel provided by libxml2 (as an integer) into a Rust enum
+  pub fn from_raw(error_level: bindings::xmlErrorLevel) -> XmlErrorLevel {
+    match error_level {
+      bindings::xmlErrorLevel_XML_ERR_NONE => XmlErrorLevel::None,
+      bindings::xmlErrorLevel_XML_ERR_WARNING => XmlErrorLevel::Warning,
+      bindings::xmlErrorLevel_XML_ERR_ERROR => XmlErrorLevel::Error,
+      bindings::xmlErrorLevel_XML_ERR_FATAL => XmlErrorLevel::Fatal,
+      _ => unreachable!("Should never receive an error level not in the range 0..=3"),
+    }
   }
+}
 
-  /// Human-readable informative error message
-  pub fn message(&self) -> &str {
-    let msg = unsafe { CStr::from_ptr((*self.0).message) };
+/// Wrapper around xmlErrorPtr.
+/// Some fields have been omitted for simplicity/safety
+#[derive(Debug)]
+pub struct StructuredError {
+  /// Human-friendly error message, lossily converted into UTF-8 from the underlying
+  /// C string. May be `None` if an error message is not provided by libxml2.
+  pub message: Option<String>,
+  /// The error's level
+  pub level: XmlErrorLevel,
+  /// The filename, lossily converted into UTF-8 from the underlying C string.
+  /// May be `None` if a filename is not provided by libxml2, such as when validating
+  /// an XML document stored entirely in memory.
+  pub filename: Option<String>,
+  /// The linenumber, or None if not applicable.
+  pub line: Option<c_int>,
+  /// The column where the error is present, or None if not applicable.
+  pub col: Option<c_int>,
+
+  /// The module that the error came from. See libxml's xmlErrorDomain enum.
+  pub domain: c_int,
+  /// The variety of error. See libxml's xmlParserErrors enum.
+  pub code: c_int,
+}
 
-    msg.to_str().unwrap()
+impl StructuredError {
+  /// Copies the error information stored at `error_ptr` into a new `StructuredError`
+  /// 
+  /// # Safety
+  /// This function must be given a pointer to a valid `xmlError` struct. Typically, you
+  /// will acquire such a pointer by implementing one of a number of callbacks
+  /// defined in libXml which are provided an `xmlError` as an argument.
+  /// 
+  /// This function copies data from the memory `error_ptr` but does not deallocate
+  /// the error. Depending on the context in which this function is used, you may
+  /// need to take additional steps to avoid a memory leak.
+  pub unsafe fn from_raw(error_ptr: *mut bindings::xmlError) -> Self {
+    let error = *error_ptr;
+    let message = StructuredError::ptr_to_string(error.message);
+    let level = XmlErrorLevel::from_raw(error.level);
+    let filename = StructuredError::ptr_to_string(error.file);
+
+    let line = if error.line == 0 {
+      None
+    } else {
+      Some(error.line)
+    };
```

Comment:
This might be too much but I leave you a more idiomatic way of doing these assignments

```suggestion
    let line = match error.line {
      0 => None,
      _ => Some(error.line)
    };
```